### PR TITLE
Refactor rulemanager

### PR DESCRIPTION
**Pull request by chenjiandongx**
### Describe what this PR does / why we need it

mention in #112 

### Does this pull request fix one issue?

#112 

### Special notes for reviews

Not ready for review, please review #109 first

### CODE REVIEW

**Comment by chenjiandongx** [comment]
Not ready for review yet.

**Comment by louyuting** [comment]
@chenjiandongx  Hi, could you please resolve the conflict?

**Comment by chenjiandongx** [comment]
@louyuting conflict has been resolved and all the tests passed

File: core/flow/rule_manager.go [hunk 1]
```diff
@@ -3,42 +3,154 @@ package flow
...
-	"sync"
 )
 
-// const
 var (
 	logger = logging.GetDefaultLogger()
 )
 
+// SafeTrafficControllers represents the threadSafe map storage for TrafficShapingController.
+type SafeTrafficControllers struct {
```
Comment (louyuting) [inline core/flow/rule_manager.go]: Is it better to rename to ConcurrentTrafficControllers?

File: core/flow/rule_manager.go [hunk 2]
```diff
@@ -3,42 +3,154 @@ package flow
...
+	controller map[string][]*TrafficShapingController
+	mux        sync.Mutex
```
Comment (louyuting) [inline core/flow/rule_manager.go]: RWMutex is better.
Here is a typical scenario of reading too much and writing too little.

File: core/flow/rule_manager.go [hunk 3]
```diff
@@ -3,42 +3,154 @@ package flow
...
+}
+
+// NewSafeTrafficControllers returns a new SafeTrafficControllers instance.
+func NewSafeTrafficControllers() *SafeTrafficControllers {
```
Comment (louyuting) [inline core/flow/rule_manager.go]: Expose the function "NewSafeTrafficControllers" is not good practice.

File: core/flow/rule_manager.go (hunk 1 above)
Comment (chenjiandongx) [reply core/flow/rule_manager.go]: That sounds better.

File: core/flow/rule_manager.go (hunk 2 above)
Comment (chenjiandongx) [reply core/flow/rule_manager.go]: ok, will use RWMutex

File: core/flow/rule_manager.go (hunk 3 above)
Comment (chenjiandongx) [reply core/flow/rule_manager.go]: but "SafeTrafficControllers" itself is exposed...

File: core/flow/rule_manager.go [hunk 4]
```diff
@@ -3,42 +3,154 @@ package flow
...
+	s.controller = m
+}
+
 // TrafficControllerGenFunc represents the TrafficShapingController generator function of a specific control behavior.
 type TrafficControllerGenFunc func(*FlowRule) *TrafficShapingController
 
-// TrafficControllerMap represents the map storage for TrafficShapingController.
-type TrafficControllerMap map[string][]*TrafficShapingController
+// SafeTrafficGenFuncs represents the threadSafe map storage for TrafficControllerGenFunc.
+type SafeTrafficGenFuncs struct {
```
Comment (louyuting) [inline core/flow/rule_manager.go]: Is it better to rename to concurrentTrafficGenFuncs?

File: core/flow/rule_manager.go [hunk 5]
```diff
@@ -3,42 +3,154 @@ package flow
...
+	funcs map[ControlBehavior]TrafficControllerGenFunc
+	mux   sync.Mutex
```
Comment (louyuting) [inline core/flow/rule_manager.go]: sync.RWMutex is better.

File: core/flow/rule_manager.go [hunk 6]
```diff
@@ -3,42 +3,154 @@ package flow
...
+}
+
+// NewSafeTrafficGenFuncs returns a new SafeTrafficGenFuncs instance.
+func NewSafeTrafficGenFuncs() *SafeTrafficGenFuncs {
```
Comment (louyuting) [inline core/flow/rule_manager.go]: Expose NewSafeTrafficGenFuncs is not good practice.

File: core/flow/rule_manager.go [hunk 7]
```diff
@@ -3,42 +3,154 @@ package flow
...
+	v, ok := s.funcs[key]
+	return v, ok
+}
 
 var (
-	tcGenFuncMap = make(map[ControlBehavior]TrafficControllerGenFunc)
-	tcMap        = make(TrafficControllerMap)
+	tcGenFuncMap = NewSafeTrafficGenFuncs()
+	tcMap        = NewSafeTrafficControllers()
 	tcMux        = new(sync.RWMutex)
```
Comment (louyuting) [inline core/flow/rule_manager.go]: tcMux could be deleted.

**Comment by louyuting** [comment]
Could you please take a look at my comment on #112 
https://github.com/alibaba/sentinel-golang/issues/112#issuecomment-603321126

**Comment by louyuting** [comment]
It seems over design, I will close this PR.
Thanks.

**Comment by chenjiandongx** [comment]
@louyuting It's okay for me :), but there are some discussions left in #112 .

**Pull request closed without merging**